storm-core/src/clj/org/apache/storm/daemon/executor.clj

```diff
@@ -446,7 +451,7 @@
     (.ack spout msg-id)
     (task/apply-hooks (:user-context task-data) .spoutAck (SpoutAckInfo. msg-id task-id time-delta))
     (when time-delta
-      (stats/spout-acked-tuple! (:stats executor-data) (:stream tuple-info) time-delta))))
+      (stats/spout-acked-tuple! (:stats executor-data) (StormMetricRegistry/counter "acked" (:worker-context executor-data) (:component-id executor-data) (pr-str (:executor-id executor-data)) (:stream tuple-info)) (:stream tuple-info) time-delta))))
```

Comment:
nit: how expensive is it to get the counter?

Building up the metric name does not sound ideal, and this is on the critical path.
 
This goes for a lot of places.

Reply:
It's not ideal. Here's some background from earlier review:

@HeartSaVioR:
>> Looks like we compose metric name and lookup from REGISTRY every time, even without executor ID and stream ID. I can see more calculation should be done after addressing, but not sure how much it affects performance. If we could also memorize metric name per combination of parameters it might help, but I'm also not sure how much it will help.

@ptgoetz:
>Prior to adding stream ID, we could store the metric as a variable and reuse it without having to do a lookup on each use. Adding stream ID required (unless I'm missing something) doing the lookup on every use. There might be additional optimizations, but because metrics names are composed of several fields, some level of string concatenation is unavoidable. For example, we could try to optimize lookups by caching metrics with metric name as the key, but we would still have to do concatenation to create the lookup key.

>I suppose we could create a MetricName class to serve as the cache key, but we'd be trading string concatenation for object creation (unless we name MetricName mutable with setX methods).

Reply:
I think we can turn it into just a map lookup, although I don't know how much work that will be.

We should know the stream names ahead of time, from the StormTopology.  As such we can create all of the metrics ahead of time for each bolt/spout instance, and cache them in a map keyed by the stream id.  Then when a tuple is Emitted/Acked/etc. we look up the Counter based off of the stream id and use it.  No object allocation and no string concatenation involved.

If we don't want to bother with parsing the StormTopology and prepopulating it, we could do lazy initialization like we are doing now.

Reply:
@revans2 regarding map lookup, wouldn't we have to concatenate in order to create the lookup key?

Reply:
I was thinking that the map would have just the stream id as the key, but the name of the metric could be anything.  i.e.

```
public class TaskStats {
   ConcurrentMap<String, Counter> ackByStream = ...;
   TaskStats(String topoId, String componentId, int taskid, int workerPort) {
        ///Save these away...
   }

   public Counter getAck(String streamId) {
      Counter c = ackByStream.get(streamId);
       if (c == null) {
           c = StormMetricRegistry.counter("acked", topoId, componentId, taskId, streamId);
           ackByStream.put(streamId, c);
       }
       return c;
   }
}
```

We could then have a map of TaskStats in the executor-data, one for each executorId the executor is over.  Or like I said if you want to go crazy you could take the list of known streams (inbound and outbound depending on the stat) and populate the metrics in the constructor, but it is not necessary.

Reply:
That makes a lot more sense now. Thanks for clarifying.